Display Input Code Excerpts in Error Messages

stu204767 opened the pull request:
### Issue
Closes #36.

### Description of the Change
* An Effect allowing access to the file path and content of an input file is added.
* Types and type classes allowing access of the line and column data for both `haskell-src-exts` and `ghc-lib-parser` source spans are added.
* Functions using source span and input file data to display excerpts of the input code in an user-friendly way are added.
* The transformation functions in the `HST.Frontend.HSE.From` and `HST.Frontend.GHC.From` modules are adjusted to display these input excerpts for unsupported language constructs.
* Input excerpts are also displayed for many other errors occuring during the pattern matching compilation. (This point may be covered by a different pull request.)
* The input code excerpts are displayed in the following way:
* Single-line excerpts:
```
Error: Enumerations are not supported!
head.hs:8:8-8:16:
8 | test = [1 .. 5] ++ [6]
           ^^^^^^^^
```
* Multi-line excerpts:
```
Error: Enumerations are not supported!
head.hs:8:8-10:16:
            vvvvvvvv
 8 | test = [1
 9 |           ..
10 |              5] ++ [6]
            ^^^^^^^^
```
```
Error: Enumerations are not supported!
head.hs:8:8-11:11:
            vvvvv
 8 | test = [1
 9 |           ..
10 |    5
11 |          ] ++ [6]
        ^^^^^^^
```

### Alternate Designs
Some alternate designs are mentioned in #36. These designs have been discussed in the meeting on Friday, in which the proposed design has been agreed on.

### Verification Process
I tested the code excerpts with some not supported constructs (the examples above and some more), different file paths and with both front ends.

### Additional Notes
* In order to not clutter the output, code excerpts are currently only shown for fatal errors, i. e. not for `skipNotSupported`. I think we should implement some kind of option for enabling or disabling messages for these skipped parts.
* Initially, I wanted colorize the spanned source code, but since not all terminals support ANSI escape code sequences directly and I didn't want to use an IO monad for that (see [here](https://hackage.haskell.org/package/ansi-terminal)), I ultimately chose against colorized code.
* Unlike single-line excerpts, which should be fairly easy to understand (I chose a representation very similar to regular Haskell error messages), multi-line excerpts are a bit more comlicated. To be as compact as possible while still marking the exact start and end of the spanned code, I chose to add a line above and below the excerpt:
  * In the upper line, the symbols (`v`) start with the beginning of the spanned code and end with the end of the longest line contained in the code excerpt, where the last line counts until the end of the spanned code.
  * In the lower line, the symbols (`^`) start with the first non-space symbol of any line contained in the code excerpt, where the first line counts from the beginning of the spanned code, and end with the end of the spanned code.

stu204767 committed:
Add effect for storing and accessing the input source code #36

stu204767 committed:
Simplify and add comments to `InputFile.hs` #36

stu204767 committed:
Add a new type for source spans and a corresponding type class #36
The new type can be used to display excerpts of an input file.
The type class includes a function for transforming a source span to this new type.

stu204767 committed:
Add `SimpleLoc` instance for `haskell-src-exts` #36

stu204767 committed:
Add `SimpleLoc` instance for `ghc-lib-parser` #36

stu204767 committed:
Format code #36

stu204767 committed:
Update `haskell-src-transformations.cabal` #36

just95 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -438,3 +441,29 @@ instance Eq (SrcSpan a) where
+  , endLine     :: Int
+  , endColumn   :: Int
+  }
+ deriving ( Eq, Show )
```
Why don't we just make this part of the `SrcSpan` data type, i.e., something like the following?

```haskell
data SrcSpan a
  = SrcSpan { srcSpanStartLine   :: Int
            , srcSpanStartColumn :: Int
            , srcSpanEndLine     :: Int
            , srcSpanEndColumn   :: Int
            , originalSrcSpan    :: SrcSpanType a
            }
  | NoSrcSpan
```

This would make the code much simpler since no type classes and wrapper data types are required for the conversion.

Does the source span of GHC-lib contain the file name? If it does, it should be part of the source span. Once we add support for the module system (see #11) we cannot be sure that the source span belongs to the current input file.

By convention, the names of the fields should contain `srcSpan`.

just95 commented on src/lib/HST/Effect/InputFile.hs:
```diff
@@ -0,0 +1,51 @@
+  ( -- * Effect
+    InputFile
+    -- * Actions
+  , setFile
+  , getFile
```
I think that the names should be as follows.

```suggestion
  , setInputFile
  , getInputFile
```

just95 commented on src/lib/HST/Effect/InputFile.hs:
```diff
@@ -0,0 +1,51 @@
+    :: Sem (InputFile ': r) a -> Sem (State (FilePath, [String]) ': r) a
+  inputFileToState = reinterpret \case
+    SetFile path content -> put (path, lines content)
+    GetFile              -> get
```
Make sure to [end every file with a new line](https://github.com/FreeProving/guidelines/blob/main/CONTRIBUTING.md#include-newline-at-end-of-file). Same in `src/lib/HST/Frontend/HSE/Config.hs` and `src/lib/HST/Frontend/Syntax.hs`.

just95 commented on src/lib/HST/Effect/InputFile.hs:
```diff
@@ -0,0 +1,51 @@
+--   the input file.
+--
+--   While setting the input file, the content is converted to lines of code.
+data InputFile m a where
+  SetFile :: FilePath -> String -> InputFile m ()
+  GetFile :: InputFile m (FilePath, [String])
```
If the file name is added to the source span, it would make sense to generalize this effect such that there can be multiple input files.

```suggestion
  GetFile :: FilePath -> InputFile m [String]
```

This would mean that a map has to be stored by `runWithInputFile` but would provide a cleaner default value for `runInputFile`.

just95 replied:
I'm not sure whether a reader effect may not be more appropriate in this situation, i.e., remove the `SetFile` action. This would also be more flexible since we could add an alternative handler that reads the file using `IO` (plus caching).

stu204767 committed:
Use a map allowing to store multiple input files in the input file effect #36

stu204767 committed:
Move additional source span information for simplification #36

stu204767 committed:
Apply syntax changes to the HSE front end #36

stu204767 committed:
Apply syntax changes to the GHC front end #36

stu204767 committed:
Format code #36

stu204767 committed:
Apply HLint suggestion #36

stu204767 committed:
Add message building functions for code excerpts #36

stu204767 committed:
Run the input file effect in `Main.hs` #36

stu204767 committed:
Use input code excerpts for critical errors with the HSE front end #36

stu204767 committed:
Use input code excerpts for critical errors with the GHC front end #36

stu204767 committed:
Resolve type errors #36

stu204767 committed:
Resolve type errors in tests #36

stu204767 committed:
Format code #36

stu204767 committed:
Apply HLint suggestion #36

stu204146-kiel commented on src/exe/Main.hs:
```diff
@@ -99,11 +101,13 @@ application = do
 processInputFile inputFilename = do
   input <- embed $ readFile inputFilename
   frontend <- parseFrontend =<< getOpt optFrontend
+  _ <- addInputFile inputFilename input
```
Why did you add a wildcard?

stu204767 replied:
Because I remembered a GHC warning I once got for not explicitly assigning such a return value to wildcard. But the GHC probably won't complain in this case, so I will change that.

stu204146-kiel commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+    Nothing      ->
+      "No content was found for the file `" ++ S.srcSpanFilePath src ++ "`!"
+    Just content ->
+      let ls = getLines (S.srcSpanStartLine src - 1) (S.srcSpanEndLine src)
```
I think it would be better to put the `-1` into the function `getLines` because the `-1` after `S.srcSpanStartLine src` can be confusing.

stu204767 replied:
I wanted `getLines` to be a generic sublist function, so I like it better the way it is now. And I don't think it is that confusing, there are `1`s added to or subtracted from these source span values at many places in this function and its subfunctions.

stu204146-kiel commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+  --   exclusive. Invalid indices do not cause runtime errors.
+  --   Example: getLines 1 3 [1, 2, 3, 4, 5] = [2, 3]
+  getLines :: Int -> Int -> [a] -> [a]
+  getLines i1 i2 ls = (take (i2 - i1) . drop i1) ls
```
Here should be an eta-reduction.
```suggestion
  getLines i1 i2  = take (i2 - i1) . drop i1
```

stu204146-kiel commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+
+-- | Reports a fatal error that the given feature is not supported. Displays an
+--   excerpt of the input code specified by the given source span.
+notSupportedWithExcerpt
```
Are the old functions still used?

stu204767 replied:
You mean `notSupported` and `skipNotSupported`? Yes, both are used. `notSupported` is used for very few locations where we don't have a source span or where we would display the entire module. `skipNotSupported` hasn't yet been replaced by `skipNotSupportedWithExcerpt` at all, actually. Look at the pull request description for more details and give me your opinion if you want.

stu204767 replied:
The old functions have now been removed and the new functions have gotten their names.

stu204767 committed:
Apply review suggestions #36

just95 commented on src/exe/Main.hs:
```diff
@@ -99,11 +101,13 @@ application = do
 processInputFile inputFilename = do
   input <- embed $ readFile inputFilename
   frontend <- parseFrontend =<< getOpt optFrontend
+  addInputFile inputFilename input
```
The idea of removing the `setInputFile`/`addInputFile` effect entirely that I've brought up in <https://github.com/FreeProving/haskell-src-transformations/pull/45#discussion_r475528988> would allow us to move `embed . readFile` into the `runInputFile` handler.

```suggestion
  input <- getInputFile inputFilename
  frontend <- parseFrontend =<< getOpt optFrontend
```

The benifit of this approach is that we could write tests for `processInputFile` that don't actually perform any IO by simply providing an alternative handler of the `InputFile` effect.

stu204767 replied:
I've implemented this in a63367d, but there are still some IO operations left in `processInputFile` (output to file or console).

just95 commented on src/lib/HST/Effect/InputFile.hs:
```diff
@@ -0,0 +1,53 @@
+    :: Sem (InputFile ': r) a -> Sem (State (Map FilePath [String]) ': r) a
+  inputFileToState = reinterpret \case
+    AddInputFile path content -> get >>= put . Map.insert path (lines content)
+    GetInputFile path         -> Map.lookup path <$> get
```
There are also [`modify`](https://hackage.haskell.org/package/polysemy-1.3.0.0/docs/Polysemy-State.html#v:modify) and [`gets`](https://hackage.haskell.org/package/polysemy-1.3.0.0/docs/Polysemy-State.html#v:gets) functions for Polysemy's `State` effect.

```suggestion
    AddInputFile path content -> modify $ Map.insert path (lines content)
    GetInputFile path         -> gets $ Map.lookup path
```

just95 commented on src/test/HST/ApplicationTests.hs:
```diff
@@ -27,23 +28,23 @@ import qualified HST.Frontend.Syntax       as S
+            S.EqAST f
+            => Sem '[WithFrontend f, GetOpt, InputFile, Cancel, Report,
+  SetExpectation, Embed IO] ())
+        -> IO ()
```
This formatting does not look right. I'll add this to the list of bugs that we have to report to the Floskell developers.

just95 commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+  -> Sem r b
+notSupportedWithExcerpt feature srcSpan = do
+  excerpt <- displayCodeExcerpt srcSpan
+  reportFatal $ Message Error $ feature ++ " are not supported!\n" ++ excerpt
```
I think that the producer of the error message should not be responsible for formatting the excerpt. Instead the handler of the `Report` effect should be responsible for this.

```suggestion
  reportFatal $ Message srcSpan Error $ feature ++ " are not supported!"
```

This approach has multiple benefits.

 1. The message text can be compared to messages that are expected by a unit test more easily.
 2. The `Member InputFile r` constraint does not have to be passed through the program explicitly.
 3. The `notSupportedWithExcerpt` and `notSupported` functions can be unified (resolves <https://github.com/FreeProving/haskell-src-transformations/pull/45#discussion_r477294818>). If there is no code excerpt, no `S.NoSrcSpan` can be provided instead.
 4. Whenever a message is reported, we are forced (on type level) to consider whether there is a good except to show to the user.

stu204767 replied:
This has now been implemented, but it required a lot more effort than expected. And I'm still unsure about a few things (see the `TODO` comments).

just95 commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+  --   maximum length is reached.
+  padLeft :: Int -> String -> String
+  padLeft maxLength s = replicate (maxLength - length s) ' ' ++ s
+{- TODO Unfinished. Could be used to colorize the spanned code.
+  insertColor :: Int -> Int -> [String] -> Maybe [String]
+  insertColor i1 i2 ls = reverse (insertColor i2 (reverse insertColor' i1 ls)))
+
+  insertColor' :: Int -> [String] -> Maybe [String]
+  insertColor' c (h:ls') = 
+  insertColor' _ [] = Nothing
+
+  insertAt :: Int -> a -> [a] -> Maybe [a]
+  insertAt 0 e list = Just (e : list)
+  insertAt i e [] = Nothing
+  insertAt i e (h:list') | i < 0 = Nothing
+                         | otherwise = fmap (h :) (insertAt (i - 1) e list')
+  -}
```
Colorization of error messages is a good idea. However, we should probably use a dedicated library for the insertion of ANSI escape codes. Let's tackle this in a separate issue.

stu204767 replied:
I've written something about this in the pull request description and I've also linked such a library, which I have looked a bit further into. If we want to support older Windows Terminals (older than the Windows 10 version of November 2015), we would need to use an IO monad everywhere where we want to print out code excerpts.

just95 replied:
What I actually wanted to say is that we should either create an issue if we want to look into this or delete the comment.

```suggestion
```

stu204767 replied:
Okay, I've deleted the comment for now, but we could still decide to use colorized text later.

just95 commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+  maybeContent <- getInputFile (S.srcSpanFilePath src)
+  return $ case maybeContent of
+    Nothing      ->
+      "No content was found for the file `" ++ S.srcSpanFilePath src ++ "`!"
```
Since there is nothing the user can do about this, we can just ignore this silently and print no excerpt in this case and when the source span is invalid.

just95 commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+  --   exclusive. Invalid indices do not cause runtime errors.
+  --   Example: getLines 1 3 [1, 2, 3, 4, 5] = [2, 3]
+  getLines :: Int -> Int -> [a] -> [a]
+  getLines i1 i2 = take (i2 - i1) . drop i1
```
Shouldn't this be equivalent to the following?

```suggestion
  getLines i1 i2 = drop i1 . take i2
```

stu204767 replied:
Yes, but according to [this](https://stackoverflow.com/questions/8529814/get-a-sublist-in-haskell), the current order may be a bit more efficient.

just95 commented on src/lib/HST/Frontend/Transformer/Messages.hs:
```diff
@@ -24,3 +26,142 @@ skipNotSupported
+    : show (S.srcSpanStartColumn src)
+    ++ '-' : show (S.srcSpanEndLine src) ++ ':' : show (S.srcSpanEndColumn src)
+
+  -- | Adds line numbers to each given line and adds marks showing the start
+  --   and end of the spanned code.
+  prettyLines :: [String] -> [String]
+  prettyLines ls
```
I'm not sure how helpful it is to show all lines of multi-line source spans. The GHC deals with multi-line source spans by showing only the first line. The `^^^` markers are followed by an ellipsis in this case to indicate that the source span has been truncated. 

```
Error: Enumerations are not supported!
head.hs:8:8-10:16:
   |
 8 | test = [1
   |        ^^...
```

Of course this is not perfect as the relevant information may be hiding in the truncated lines. However, when all lines are shown, the actual error message gets obfuscated. Since the text of the error message is more important than the code excerpt (which the user can simply lookup in their editor), I'd recommend to adopt the GHC's approach.

stu204767 replied:
This is not a big problem currently, because these excerpts are currently only shown for fatal errors, so we have a maximum of one excerpt per program call. I've written something about this in the pull request description, so please read it.
And while the GHC does display only one regular line of the input code, it practically displays much more in many cases describing the location of the error (I have constructed an example where it prints out as much as I want). I definitely think that we should display more than one line, but we could set a maximum in some way. One idea I have is to display up to five lines and to display the first two and last two lines of longer excerpts, like this:
```
head.hs:8:8-101:11:
             vvvvv
  8 | test = [1
  9 |           ..
  ⁝
100 |    5
101 |          ] ++ [6]
         ^^^^^^^
```

just95 replied:
> This is not a big problem currently, because these excerpts are currently only shown for fatal errors

If multi-line code excerpts prevent us from having code excerpts in non-fatal messages, we do have a big problem :D Especially in case of non-fatal messages the user needs enough information to quickly decide whether they want to investigate the warning but not too much information that they miss the warning altogether.

> the GHC [...] practically displays much more in many cases describing the location of the error

While I am aware that the GHC often prints additional code (using it's pretty printer) in messages like "In a stmt of a 'do' block: …", the actual code excerpts (i.e., ["caret diagnostics"](https://hackage.haskell.org/package/ghc-8.10.1/docs/src/ErrUtils.html#getCaretDiagnostic)) include the first line only as far as I know.

> One idea I have is to display up to five lines and to display the first two and last two lines of longer excerpts

That would be a viable option. Alternatively, we could show the first line by default and add a command line flag to show multi-line excerpts.

stu204767 replied:
> If multi-line code excerpts prevent us from having code excerpts in non-fatal messages, we do have a big problem :D

Multi-line excerpts are not the reason I didn't implement code excerpts for `skipNotSupported` yet. I think we should offer an option to disable these warnings altogether.

stu204767 committed:
Use the reader effect instead of the state effect in `InputFile.hs` #36

stu204767 committed:
Add abbreviation for excerpts containing more than five lines #36

stu204767 committed:
Improve markings of excerpts #36

stu204767 committed:
Apply further review suggestions #36

stu204767 committed:
Move the display of code excerpts to the Report effect (draft, errors exist) #36

stu204767 committed:
Move basic source span data to the new type `MsgSrcSpan` #36

stu204767 committed:
Resolve type errors by using the new `MsgSrcSpan` type and a new `message` constructor #36

stu204767 committed:
Resolve type errors in the `HST.Frontend.Transformer.Messages` module #36

stu204767 committed:
Adapt the HSE front end modules to the message reporting changes #36

stu204767 committed:
Adapt the GHC front end modules to the message reporting changes #36

stu204767 committed:
Adapt remaining modules to the message reporting changes #36

stu204767 committed:
Adapt the `Main` module to the changes of the `Report` effect #36

stu204767 committed:
Add additional pretty-printing function for messages #36
This new function does not require the `InputFile` effect and is currently only used for testing purposes

stu204767 committed:
Remove unnecessary `InputFile` constraint #36

stu204767 committed:
Resolve type errors in tests #36

stu204767 committed:
Adapt `ReportTests.hs` to changes in the report effect #36

stu204767 committed:
Format code #36

stu204767 committed:
Make some unqualified imports in the test modules explicit #36

stu204767 committed:
Display code excerpts for skipped unsupported constructs as well #36

stu204767 committed:
Remove `notSupported` functions without code excerpts #36

stu204767 committed:
Format code #36

just95 commented on src/lib/HST/Effect/Report.hs:
```diff
@@ -54,13 +58,161 @@ import           HST.Effect.Cancel ( Cancel, cancel, runCancel )
+  ++ '-'
+  : show (S.msgSrcSpanEndLine src) ++ ':' : show (S.msgSrcSpanEndColumn src)
+
+-- TODO Should the following function be in its own module? It can't be put
+-- back into 'HST.Frontend.Transformer.Messages' because of cyclic imports.
```
We could move the entire code for messages into its own module.

stu204767 replied:
I think this new module should include the functions in `HST.Frontend.Transformer.Messages` and be located in the `Util` folder. What do you think?

just95 replied:
The `Util` folder seems to be a good choice for the new module but I would recommend to keep the modules separate. Otherwise there is a cyclic dependency since the `HST.Frontend.Transformer.Messages` module uses the `Report` effect but `HST.Effect.Report` needs to import the new module.

just95 commented on src/exe/Main.hs:
```diff
@@ -88,7 +89,11 @@ application = do
         then embed putUsageInfo
-        else runInputFile $ mapM_ processInputFile inputFiles
+        else 
+          -- TODO There probably is a better way to add actual input files to the
+          -- Report effect
+          runInputFile inputFiles . reportToHandleOrCancel stderr
+          $ mapM_ processInputFile inputFiles
```
Why does the handler need to know all input files in advance? Wouldn't it be possible to load the input files on demand by doing something like the following.

```haskell
maybeContents <- gets $ Map.lookup path
case maybeContents of
  Nothing -> do
    contents <- embed $ readFile path
    modify $ Map.insert path contents
    return contents
  Just contents -> return contents
```

I think this should allow us to invoke the `runInputFile` and `reportToHandleOrCancel` handlers just in `main`.

stu204767 replied:
That's the reason I didn't change the `InputFile` effect to use [Polysemy's `Reader` effect](https://hackage.haskell.org/package/polysemy-1.3.0.0/docs/Polysemy-Reader.html) initially, but used the `State` effect with an `addInputFile` method.
Can this somehow be implemented with the `Reader` effect (possibly with the unexplained `Local` action)? I thought (not) being able to change the state of the effect was the main difference between the `State` and `Reader` effect.
I can of course change `InputFile` to use the `State` effect once again.

just95 replied:
You can use Polysemy's `State` effect internally instead of the `Reader` effect even if the `InputFile` effect itself appears like a reader effect to the user since no equivalent to the `put` operation is exposed.

stu204146-kiel commented on src/lib/HST/Effect/InputFile.hs:
```diff
@@ -0,0 +1,52 @@
+-------------------------------------------------------------------------------
+-- | Handles a computation by reading the given files and providing a file map
+--   with their file paths and contents.
+runInputFile
```
I think we should call the function `runInputFiles` because there can be multiple files in the given list.

stu204767 replied:
If we rename that function, we should also rename the effect and the module itself to `InputFiles`, in my opinion.
I can do that, if you want that, though I would personally leave the name as it is.

stu204146-kiel commented on src/lib/HST/Effect/Report.hs:
```diff
@@ -54,13 +58,161 @@ import           HST.Effect.Cancel ( Cancel, cancel, runCancel )
+--   source span numbers, all lines of the input program that are at least
+--   partially contained in the given source span, including their line
+--   numbers, and marks showing the start and end of the spanned code.
+displayCodeExcerpt :: Member InputFile r => Maybe S.MsgSrcSpan -> Sem r String
```
In my tests, I found out that `displayCodeExcerpt` can also display the following `SrcSpan` that is invalid.
`S.MsgSrcSpan "./example/ExampleQueue1.hs" 1 4 1 3`.

stu204767 replied:
This should be fixed now, but you should run your tests again to be sure that it's actually fixed and that I haven't introduced another bug.

just95 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -438,3 +442,27 @@ instance Eq (SrcSpan a) where
 -- | Custom order for 'SrcSpan' which treats all source spans as equal.
 instance Ord (SrcSpan a) where
   _ `compare` _ = EQ
+
+-- | Type for storing basic data of a source span, i. e. the file path and the
+--   line and column of the start and end of the spanned source code.
+--
+--   The data contained in this type is used for displaying input code
+--   excerpts. This additional type is necessary in order to have a type for
+--   source spans without type variables, which would cause problems in the
+--   "HST.Effect.Report" module.
+data MsgSrcSpan = MsgSrcSpan
+  { msgSrcSpanFilePath    :: FilePath
+  , msgSrcSpanStartLine   :: Int
+  , msgSrcSpanStartColumn :: Int
+  , msgSrcSpanEndLine     :: Int
+  , msgSrcSpanEndColumn   :: Int
+  }
+ deriving ( Eq, Show )
```
I think the only reason why the `MsgSrcSpan` data type cannot be in the same module as the remaining message code is that the `message` smart constructor takes a `S.SrcSpan`. By adding a type class for `toMsgSrcSpan`, this dependency could be eliminated.

stu204767 replied:
I would leave the `MsgSrcSpan` in this location, because I like that the `Syntax` module currently has no imports and stores the entirety of our ASTs in types defined in the module itself, except for the types represented by the type families. And I also think that `MsgSrcSpan` fits into the `Syntax` module: It was only created because there were problems with having a type variable in the `SrcSpan` type and I don't think this solution to these problems should cause the module these source span data are stored in to be changed.
However, if you really want to have `MsgSrcSpan` in the new `Message` module, I will move it.

stu204767 committed:
Fix a bug where invalid source span were seen as valid #36

stu204767 committed:
Move message types and functions to a new module #36

stu204767 committed:
Change the `InputFile` effect to use Polysemy's `State` effect internally #36

stu204767 committed:
Link `report` function for Haddock #36

stu204767 committed:
Merge branch 'main' into issue-36

stu204767 committed:
Resolve errors that emerged due to merging #36

stu204767 committed:
Format code #36
